docs/root/configuration/http/http_filters/api_key_auth_filter.rst

```diff
@@ -0,0 +1,120 @@
+.. _config_http_filters_api_key_auth:
+
+API key auth
+============
+
+This HTTP filter can be used to authenticate users based on the unique API key. The filter will
+extract the API keys from either an HTTP header, a parameter query, or a cookie and verify them against
+the configured credential list.
+
+If the API key is valid and the related client is allowed, the request will be allowed to continue.
+If the API key is invalid or not exists, the request will be denied with 401 status code.
+If the API key is valid but the related client is not allowed, the request will be denied with
+403 status code.
+
+Configuration
+-------------
+
+* This filter should be configured with the type URL ``type.googleapis.com/envoy.extensions.filters.http.api_key_auth.v3.ApiKeyAuth``.
+* :ref:`v3 API reference <envoy_v3_api_msg_extensions.filters.http.api_key_auth.v3.ApiKeyAuth>`
+
+An example configuration of the filter may look like the following:
+
+.. code-block:: yaml
+
+  http_filters:
+  - name: api_key_auth
+    typed_config:
+      "@type": type.googleapis.com/envoy.extensions.filters.http.api_key_auth.v3.ApiKeyAuth
+      credentials:
+        entries:
+        - api_key: one_key
+          client_id: one_client
```

Comment:
Probably a dumb question: where does the "one_client" value coming from in the data-plane (how it the "related client" defined)?

Reply:
> Probably a dumb question: where does the "one_client" value coming from in the data-plane (how it the "related client" defined)?

It's hard to answer this quesion. the `client_id` is an identity that related to the `key`. It could be anything that could be used to identify a client/app or a group of clients/apps.

For a specific example, the gateway production may provides the ability to manage the credentials. And that credentials management module may allow users to create or define the `client` or `consumer`, and then, users could create some keys that related to the `client`/`consumer`.
In conclusion, it depends on the abstraction of the gateawy production.